optuna/storages/_journal/file.py

```diff
@@ -0,0 +1,175 @@
+                os.symlink(self._lock_target_file, self._lockfile)
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    continue
+                else:
+                    raise err
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@experimental_class("3.1.0")
+class JournalFileOpenLock(JournalFileBaseLock):
+    """Lock class for synchronizing processes for NFSv3 or later.
+
+    On acquiring the lock, open system call is called with the O_EXCL option to create an exclusive
+    file. The file is deleted when the lock is released. This class is only supported when using
+    NFSv3 or later on kernel 2.6 or later. In prior NFS environments, use
+    :class:`~optuna.storages.JournalFileSymlinkLock`.
+
+    Args:
+        filepath:
+            The path of the file whose race condition must be protected.
+    """
+
+    def __init__(self, filepath: str) -> None:
+        self._lockfile = filepath + LOCK_FILE_SUFFIX
+        self._lockrenamefile = self._lockfile + str(uuid.uuid4()) + RENAME_FILE_SUFFIX
+
+    def acquire(self) -> bool:
+        while True:
+            try:
+                open_flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
+                os.close(os.open(self._lockfile, open_flags))
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    continue
+                else:
+                    raise err
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@contextmanager
+def get_lock_file(lock_obj: JournalFileBaseLock) -> Iterator[None]:
+    lock_obj.acquire()
+    try:
+        yield
+    finally:
+        lock_obj.release()
+
+
+@experimental_class("3.1.0")
+class JournalFileStorage(BaseJournalLogStorage):
+    """File storage class for Journal log backend.
+
+    Args:
+        file_path:
+            Path of file to persist the log to.
+
+        lock_obj:
+            Lock object for process exclusivity.
+
+    """
+
+    def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = None) -> None:
+        self._file_path: str = file_path
+        self._lock = lock_obj or JournalFileSymlinkLock(self._file_path)
```

Comment:
It seems that if users specify `lock_obj`  they need to write the same path twice as follows:  
```python
storage = JournalFileStorage("/path/to/optuna-db", JournalFileSymlinkLock("/path/to/optuna-db"))
```
IMO, it's redundant and error prone, so how about refine the interface? (e.g., makes `lock_obj` a function that takes a filepath and returns an `JournalFileBaseLock` instance)

Reply:
The assumption is that there are very few users who change `lock_obj` (most cases should be fine if symlink lock is used, and we only provide a path in case it is necessary to change it), so this case, IMO need not be considered that much. 

As for the interface that specifies a factory function to generate `lock_obj`, although duplicate specification of filepath is eliminated, it may be necessary to prepare a factory function for each lock class, and it is somewhat questionable whether this would really make the user happy.

Reply:
>  it may be necessary to prepare a factory function for each lock class

Hmm, I thought that we could write the code as follows without any additional functions (because Python's constructors can be used as functions):
```python
storage = JournalFileStorage("/path/to/optuna-db", JournalFileSymlinkLock)
```

Reply:
I think this should be avoided because we don't define the constructor arguments of `JournalFileBaseLock`.  Users might implement their own file lock which does not have `file_path` as constructor argument. On such cases, we cannot use the passed `UserDefinedLock` as alternative to the default `JournalFileSymlinkLock`.

Reply:
> Users might implement their own file lock which does not have file_path as constructor argument

On such case, users can use `lambda` as follows I think:

```python
storage = JournalFileStorage("/path/to/optuna-db", lambda path: UserDefinedLock(...))
```

Honestly, I cannot imagine that there would be an implmentation class of `JournalFileBaseLock` that doesn't take the DB file path as a constructor argument (as it's a file based lock anyway). But, the design is a matter of taste and, as you said above, using a lock class other than  `JournalFileSymlinkLock` would be very rare case, so feel free to ignore my opinion and keep the implementation as-is.

Reply:
I was imagining something like `ThreadLock` in python. `ThreadLock` does not take the "resource to protect" as an constructor argument. 
Anyways, thank you for the comment!